### PR TITLE
Make dates be ReadableDateTimes in scripts

### PR DESCRIPTION
**Pull request by nik9000**
Instead of longs. If you want millis since epoch you can call `doc.date_field.value.millis`.

Relates to #22875

### CODE REVIEW

**Comment by nik9000** [comment]
@rjernst, would you like to look at this? It is the second step we discussed in https://github.com/elastic/elasticsearch/pull/22875#issuecomment-276461687.

File: core/src/main/java/org/elasticsearch/search/aggregations/support/values/ScriptDoubleValues.java [hunk 1]
```diff
@@ -47,36 +47,49 @@ public void setDocument(int docId) {
...
+            return ((Number) o).doubleValue();
+        } else if (o instanceof ReadableInstant) {
+            // Dates are exposed in scripts as ReadableDateTimes but aggregations want them to be numeric
+            return ((ReadableInstant) o).getMillis();
+        } else if (o instanceof Boolean) {
+            // We do expose boolean fields as boolean in scripts, however aggregations still expect
+            // that scripts return the same internal representation as regular fields, so boolean
+            // values in scripts need to be converted to a number, and the value formatter will
+            // make sure of using true/false in the key_as_string field
+            return ((Boolean) o).booleanValue() ? 1.0 : 0.0;
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/values/ScriptDoubleValues.java]: This is just a bonus fix, unrelated to the date change right?

File: docs/reference/migration/migrate_6_0/scripting.asciidoc [hunk 2]
```diff
@@ -5,3 +5,10 @@
 
 The groovy scripting language was deprecated in elasticsearch 5.0 and is now removed.
 Use painless instead.
+
+==== Date fields now return dates
+
+`doc.some_date_field.value` now returns `ReadableDateTime`s instead of
```
Comment (rjernst) [inline docs/reference/migration/migrate_6_0/scripting.asciidoc]: Is it possible to link `ReadableDateTime` here to the generated painless docs?

**Review by rjernst — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/search/aggregations/support/values/ScriptDoubleValues.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/values/ScriptDoubleValues.java]: Indeed - bonus to make this line up with `toLongValue`.

File: docs/reference/migration/migrate_6_0/scripting.asciidoc (hunk 2 above)
Comment (nik9000) [inline docs/reference/migration/migrate_6_0/scripting.asciidoc]: Sure!

**Review by nik9000 — COMMENTED** [review]
Given this [conversation](https://github.com/elastic/elasticsearch/issues/21121#issuecomment-277753526) I don't think we're done here but I still think this is a step in the right direction so I'm going to retest locally and merge.

**Comment by nik9000** [comment]
master: 0d6e62224209c0153eae429d22c2474c091f6990

**Pull request merged**